src/test/java/fr/sncf/osrd/infra_state/RouteStateTest.java

```diff
@@ -150,4 +155,133 @@ public void testReserveStatusChanges() throws InvalidInfraException, SimulationE
+
+        schedule.trainSchedules.forEach(s -> {
+            s.routes = (ID<RJSRoute>[]) new ID[]{
+                    new ID<RJSRoute>("rt.C3-S7"),
+                    new ID<RJSRoute>("rt.S7-buffer_stop_c"),
+            };
+            s.initialHeadLocation.trackSection = new ID<>("ne.micro.foo_to_bar");
+            s.initialHeadLocation.offset = 100;
+        });
+
+
+        var trainSchedules = RJSSimulationParser.parse(rjsInfra, schedule);
+        var config = new Config(
+                jsonConfig.simulationTimeStep,
+                rjsInfra,
+                trainSchedules,
+                null,
+                jsonConfig.simulationStepPause,
+                false,
+                jsonConfig.realTimeViewer,
+                jsonConfig.changeReplayCheck
+        );
+        var sim = Simulation.createFromInfraAndEmptySuccessions(RailJSONParser.parse(infra),
+                0, null);
+
+        run(sim, config);
+    }
+
+    @Test
+    public void testCircularInfraReserves() throws InvalidInfraException {
+        final var infra = getBaseInfra("circular_infra/infra.json");
+        final var config = getBaseConfig("circular_infra/config.json");
+
+        var changelog = new ArrayChangeLog();
+
+        var sim = Simulation.createFromInfraAndSuccessions(RailJSONParser.parse(infra),
+                config.switchSuccessions, 0, changelog);
+
+        config.trainSchedules.remove(2);
+        config.trainSchedules.remove(1);
+
+        run(sim, config);
+
+        var changesSet = changelog.publishedChanges.stream()
+                .filter(x -> x instanceof RouteState.RouteStatusChange)
+                .map(Object::toString)
+                .collect(Collectors.toSet());
+
+        // We check that every route has been reserved and occupied at least once
+        for (int i = 0; i < infra.routes.size(); i++) {
+            for (var status : new RouteStatus[]{RouteStatus.RESERVED, RouteStatus.OCCUPIED}) {
+                var expected = new RouteState.RouteStatusChange(sim, sim.infraState.getRouteState(i), status);
+                assert changesSet.contains(expected.toString());
+            }
+        }
+    }
+
+    @Test
+    @Disabled("Fixing this requires changes in the API and middle/front end, it will be done later")
+    public void testCircularInfraRouteIndexes() throws InvalidInfraException {
+        final var infra = getBaseInfra("circular_infra/infra.json");
+        final var config = getBaseConfig("circular_infra/config.json");
+
+        var changelog = new ArrayChangeLog();
+
+        var sim = Simulation.createFromInfraAndSuccessions(RailJSONParser.parse(infra),
+                config.switchSuccessions, 0, changelog);
+
+        run(sim, config);
+
+        for (var train : sim.trains.values()) {
+            var trainState = train.getLastState();
+            var path = train.schedule.plannedPath;
+            assertEquals(path.routePath.size() - 1, trainState.routeIndex);
+        }
+    }
 }
```

Comment:
I'll open an issue for this test. Fixing it requires synchronization between back, middle and front, now is not an appropriate time.